numpy/f2py/crackfortran.py

```diff
@@ -2536,6 +2536,10 @@ def analyzevars(block):
                 for k in ['public', 'private']:
                     if k in gen:
                         vars[n] = setattrspec(vars[n], k)
+            for n in [b['name'] for b in block['body'] if b['name'] not in vars]:
+                for k in ['public', 'private']:
+                    if k in gen:
+                        vars[n] = {'attrspec': [k, ]}
```

Comment:
Why not match the loop above?

```suggestion
                        vars[n] = setattrspec(vars[n], k)
```

Reply:
Because, n is not in vars yet (that's the purpose of the loop and the patch: to add missing entries in vars). So there is nothing to extend with the setattrspec() routine and I prefer then to create directly a new dictionary when setting vars[n].

Reply:
When #15781 is accepted, I can complete the tests newly introduced there with a test showing the issue.

Reply:
Perhaps worth combining the two loops into one?
```python
            all_var_names = set(vars) | set(b['name'] for b in block['body'])
            for n in all_var_names:
                for k in ['public', 'private']:
                    if k in gen:
                        vars[n] = setattrspec(vars.get(n, {}), k)
```

Reply:
very nice ;) Updating.